README.md
# Welcome to Facebook Account Kit Flutter Version!
### What's that?
As you all know, Facebook Account Kit - powered by Facebook - lets you quickly register for and login to your app by using just user's phone number. It's reliable and easy to integrate. Now it's is available in **Flutter**.

## Let's prepare some your own Facebook Developer information !!!
- Access Facebook Developer : [https://developers.facebook.com](https://developers.facebook.com/)
- Create new app
![Create new app](https://sanvay.herokuapp.com/images/screenshot_1.png)
- Add AccountKit to your Facebook Developer project
![Add AccountKit to your Facebook Developer project](https://sanvay.herokuapp.com/images/screenshot_2.png)
#### Note: If your project just provide for only 1 platform. You can skip the other part.
#### *** For Android Configuration:
- Open your project, create `strings.xml` file at this path: `\android\app\src\main\res`
- Add 2 variables like below:

```
<?xml version="1.0" encoding="utf-8"?>  
<resources>  
    <string name="FACEBOOK_APP_ID" translatable="false">[YOUR_FACEBOOK_APP_ID]</string>  
    <string name="ACCOUNT_KIT_CLIENT_TOKEN" translatable="false">[ACCOUNT_KIT_CLIENT_TOKEN]</string>  
</resources>
```

*About the Facebook App ID and Account Kit Client Token, please keep reading*

- Set up your AndroidManifest.xml in `/android/app/src/main`. Full code here.

#### *** For iOS Configuration:
- Open `ios\Runner\Info.plist` by right click -> open by source code
- Copy and paste this inside the space between **\<dict>\</dict>** tag.

```
	<key>FacebookAppID</key>
    <string>[YOUR_FACEBOOK_APP_ID]</string>
    <key>AccountKitClientToken</key>
    <string>[ACCOUNT_KIT_CLIENT_TOKEN]</string>
    <key>CFBundleURLTypes</key>
    <array>
        <dict>
            <key>CFBundleURLSchemes</key>
            <array>
                <string>ak[YOUR_FACEBOOK_APP_ID]</string>
            </array>
        </dict>
    </array>
```

#### About Facebook App Id and Account Kit Client Token:
*YOUR_FACEBOOK_APP_ID: you can get it in picture below and simply press to copy*
![YOUR_FACEBOOK_APP_ID](https://sanvay.herokuapp.com/images/screenshot_3.png)
*ACCOUNT_KIT_CLIENT_TOKEN:*
![ACCOUNT_KIT_CLIENT_TOKEN](https://sanvay.herokuapp.com/images/screenshot_4.png)

...
It's done for configuration !! Let's go back to Flutter **main.dart**

## How to use?
It's very easy !!!
### 1. Add dependency to your pubspec.yaml file:

    dependencies:  
	    fb_accountkit_flutter: ^0.0.2

And do not forget to call `flutter get` to get the dependency in **pub.dev**

### 2. Import to your code:

    import 'package:fb_accountkit_flutter/fb_accountkit_flutter.dart';


### 3. Call open AccountKit whenever you want:

    Result result = await FbAccountkitFlutter.startAuthentication()

## Example Function
```
import 'dart:async';  
import 'package:flutter/services.dart';  
import 'package:fb_accountkit_flutter/fb_accountkit_flutter.dart';

Future<void> openFacebookActivity() async {  
  Result result;  
  
  try {  
    FbAccountkitFlutter.defaultCountryCode = "[YOUR DEFAULT COUNTRY CODE]"; // Ex:  US, AL etc.  
	// Find more here : https://developers.facebook.com/docs/accountkit/countrycodes/  
	FbAccountkitFlutter.codeIso = "[YOUR ISO CODE]"; // Ex: 93, 355 etc.  
	// Find Dialing Code here : https://developers.facebook.com/docs/accountkit/countrycodes/  
	FbAccountkitFlutter.countryCode = "[YOUR COUNTRY CODE]";  
    // Find A 3 in here : http://kirste.userpage.fu-berlin.de/diverse/doc/ISO_3166.html  
	
	// Optional  
	FbAccountkitFlutter.phone = "[INITIAL PHONE]";  
	
	// Request open accountkit  
	result = await FbAccountkitFlutter.startAuthentication();  
  
  } on PlatformException catch(e) {  
    print(e);  
    result = null;  
  }  
  
}
```

You can read the full code in the example package also.


### Happy Coding !

CHANGELOG.md
## 0.0.1 
Initial release.
## 0.0.2
Ready for usage
# 1.0.0
Optimize, Clean to high. Update Document.

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:fb_accountkit_flutter_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/fb_accountkit_flutter_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const MethodChannel channel = MethodChannel('fb_accountkit_flutter');

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

}


lib/fb_accountkit_flutter.dart
import 'dart:async';
import 'package:flutter/services.dart';
import 'dart:convert';

class FbAccountkitFlutter {
  static const MethodChannel _channel = const MethodChannel('fb_accountkit_flutter');
  static const String _methodName = "open_facebook_activity";
  static const String _phoneArg = "phone";
  static const String _defaultCountryCodeArg = "default_country_code";
  static const String _countryCodeIsoArg = "country_code_iso";
  static const String _countryCodeArg = "country_code";
  static const int SUCCEED = 1;
  static const int FAILED = 0;

  static String _phone;
  static String _defaultCountryCode;
  static String _codeIso;
  static String _countryCode;

  static Future<Result> startAuthentication() async {
    Result result = Result();
    Map<String, dynamic> argument = Map();

    if (_defaultCountryCode == null) {
      result._status = FAILED;
      result._message = "Default Country Code can not be null";
      return result;
    }
    argument[_defaultCountryCodeArg] = _defaultCountryCode;

    if (_codeIso == null) {
      result._status = FAILED;
      result._message = "Code ISO can not be null";
      return result;
    }
    argument[_countryCodeIsoArg] = _codeIso;

    if (_countryCode == null) {
      result._status = FAILED;
      result._message = "Country Code can not be null";
      return result;
    }
    argument[_countryCodeArg] = _countryCode;

    if (_phone != null) {
      argument[_phoneArg] = _phone;
    }

    try {
      final String accountResult = await _channel.invokeMethod(
          _methodName, argument);
      if (accountResult == "1") {
        result._status = FAILED;
        result._message = "Facebook Authentication is Failed or Canceled";
        return result;
      }
      Result parsedAccountResult = Result.fromJson(accountResult);
      parsedAccountResult._status = SUCCEED;
      parsedAccountResult._message = "Phone Authentication Succeed !";
      return parsedAccountResult;
    } on PlatformException catch (e) {
      print(e);
      result._status = FAILED;
      result._message = e.toString();
    }
    return result;
  }

  static set phone(String value) {
    _phone = value;
  }

  static set defaultCountryCode(String value) {
    _defaultCountryCode = value;
  }

  static set codeIso(String value) {
    _codeIso = value;
  }

  static set countryCode(String value) {
    _countryCode = value;
  }
}

class Result {
  int _status;
  String _message;
  String _userId;
  String _email;
  String _phone;

  Result();

  Result.fromJson(String jsonString) {
    Map parsedJson = json.decode(jsonString);
    this._phone = parsedJson["phone"];
    this._email = parsedJson["email"];
    this._userId = parsedJson["id"];
  }

  int get status => _status;

  String get message => _message;

  String get userId => _userId;

  String get phone => _phone;

  String get email => _email;
}


ios/Classes/SwiftFbAccountkitFlutterPlugin.swift
import Flutter
import UIKit
import AccountKit

public class SwiftFbAccountkitFlutterPlugin: NSObject, FlutterPlugin, AKFViewControllerDelegate {
    let METHOD_NAME = "open_facebook_activity"
    var mResult:FlutterResult?
    var accountKit: AccountKit?
    var myViewController:UIViewController?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "fb_accountkit_flutter", binaryMessenger: registrar.messenger())
        let instance = SwiftFbAccountkitFlutterPlugin()
        instance.myViewController = (UIApplication.shared.delegate?.window??.rootViewController)!
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if call.method == METHOD_NAME {
            mResult = result
            if let arguments = call.arguments, let args = arguments as? [String: String] {
                let phone = args["phone"]
                let countryCodeIso = args["country_code_iso"]
                let countryCode = args["country_code"]
                let defaultCountryCode = args["default_country_code"]

                openFacebookAuthentication(phone: phone, defaultCountryCode: defaultCountryCode!, countryCode: countryCode!, countryISO: countryCodeIso!)
                
            }
        } else {
            print("Flutter: Null Method \(call.method)")
            result("1")
        }
    }
    
    public func openFacebookAuthentication(phone:String?, defaultCountryCode:String, countryCode:String, countryISO: String) {
        if accountKit == nil {
            accountKit = AccountKit(responseType: .accessToken)
        }
        let inputState = UUID().uuidString
        let afPhone = PhoneNumber(countryCode: countryCode, countryISO: countryISO, phoneNumber: phone ?? "")
        let vc = (accountKit?.viewControllerForPhoneLogin(with: afPhone, state: inputState))!
        vc.isSendToFacebookEnabled = true
        vc.delegate = self
        vc.defaultCountryCode = defaultCountryCode
        
        self.prepareLoginViewController(loginViewController: vc)
        myViewController?.present(vc as UIViewController, animated: true, completion: nil)
    }
    
    func prepareLoginViewController(loginViewController: AKFViewController) {
        loginViewController.delegate = self
        //UI Theming - Optional
        loginViewController.uiManager = SkinManager(skinType: .classic, primaryColor: UIColor.blue)
    }
    
    public func viewController(_ viewController: UIViewController & AKFViewController, didCompleteLoginWith accessToken: AccessToken, state: String) {
    
        let accountKit = AccountKit(responseType: .accessToken)
        accountKit.requestAccount({ (account, error) in
            if let error = error {
                print(error)
                self.mResult?("1")
                return
            }
            if let account = account, let phoneNumber = account.phoneNumber {
                let dict = [
                    "phone": phoneNumber.phoneNumber,
                    "email": account.emailAddress ?? "",
                    "id": account.accountID
                            ]
                if let jsonData = try? JSONSerialization.data(
                    withJSONObject: dict,
                    options: []) {
                    let returnedJsonString = String(data: jsonData,
                                             encoding: .ascii)
                    self.mResult?(returnedJsonString)
                }
            }
        })
    }
   
}


android/src/main/java/fb/accountkit/fb_accountkit_flutter/FbAccountkitFlutterPlugin.java
package fb.accountkit.fb_accountkit_flutter;

import android.app.Activity;
import android.content.Intent;
import android.util.Log;

import com.facebook.accountkit.Account;
import com.facebook.accountkit.AccountKit;
import com.facebook.accountkit.AccountKitCallback;
import com.facebook.accountkit.AccountKitError;
import com.facebook.accountkit.AccountKitLoginResult;
import com.facebook.accountkit.PhoneNumber;
import com.facebook.accountkit.ui.AccountKitActivity;
import com.facebook.accountkit.ui.AccountKitConfiguration;
import com.facebook.accountkit.ui.LoginType;

import org.json.JSONException;
import org.json.JSONObject;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.Registrar;

import static android.app.Activity.RESULT_OK;

/** FbAccountkitFlutterPlugin */
public class FbAccountkitFlutterPlugin implements MethodCallHandler, PluginRegistry.ActivityResultListener {
  private final Activity activity;
  private static final int FACEBOOK_LOGIN_REQUEST_CODE = 123;
  private static final String FAILED = "1";
  private static final String CHANNEL_NAME = "fb_accountkit_flutter";
  private static final String METHOD_NAME = "open_facebook_activity";
  private static final String PHONE_ARGUMENT= "phone";
  private static final String DEFAULT_COUNTRY_CODE_ARGUMENT= "default_country_code";
  private static final String COUNTRY_CODE_ISO_ARGUMENT= "country_code_iso";
  private static final String COUNTRY_CODE_ARGUMENT= "country_code";
  private MethodChannel.Result mResult;

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), CHANNEL_NAME);
    FbAccountkitFlutterPlugin flutterPlugin = new FbAccountkitFlutterPlugin(registrar.activity());
    registrar.addActivityResultListener(flutterPlugin);
    channel.setMethodCallHandler(flutterPlugin);
  }

  public FbAccountkitFlutterPlugin(Activity activity) {
    this.activity = activity;
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    mResult = result;
    switch (call.method) {
      case METHOD_NAME:
        String phone = call.argument(PHONE_ARGUMENT);
        String defaultCountryCode = call.argument(DEFAULT_COUNTRY_CODE_ARGUMENT);
        String countryCodeIso = call.argument(COUNTRY_CODE_ISO_ARGUMENT);
        String countryCode = call.argument(COUNTRY_CODE_ARGUMENT);
        openFacebookAccountKitActivity(phone, defaultCountryCode, countryCodeIso, countryCode);
        break;
      default:
        result.notImplemented();
    }
  }

  private void openFacebookAccountKitActivity(String phone, String defaultContryCode, String countryCodeIso, String countryCode) {
    Intent intent = new Intent(activity, AccountKitActivity.class);
    AccountKitConfiguration.AccountKitConfigurationBuilder configurationBuilder = new AccountKitConfiguration.AccountKitConfigurationBuilder(
            LoginType.PHONE, AccountKitActivity.ResponseType.TOKEN);
    if (defaultContryCode != null)
      configurationBuilder.setDefaultCountryCode(defaultContryCode);
    configurationBuilder.setReadPhoneStateEnabled(false);
    if (phone != null)
      configurationBuilder.setInitialPhoneNumber(new PhoneNumber(countryCodeIso, phone.substring(1), countryCode));
    intent.putExtra(AccountKitActivity.ACCOUNT_KIT_ACTIVITY_CONFIGURATION, configurationBuilder.build());
    activity.startActivityForResult(intent, FACEBOOK_LOGIN_REQUEST_CODE);

  }


  @Override
  public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
    if (requestCode == FACEBOOK_LOGIN_REQUEST_CODE && resultCode == RESULT_OK) {
      AccountKitLoginResult loginResult;
      if (data != null) {
        loginResult = data.getParcelableExtra(AccountKitLoginResult.RESULT_KEY);
        if (loginResult.getAccessToken() != null) {
          AccountKit.getCurrentAccount(new AccountKitCallback<Account>() {
            @Override
            public void onSuccess(Account account) {
              if (mResult != null) {
                ReturnedResult returnedResult = new ReturnedResult(account.getId(), account.getPhoneNumber().getPhoneNumber(), account.getEmail());
                mResult.success(returnedResult.toJson());
              }
            }

            @Override
            public void onError(AccountKitError accountKitError) {
              if (mResult != null) {
                mResult.error(FAILED, null, null);
              }
            }
          });
        }
      }
    }
    return false;
  }

  private class ReturnedResult {
    String id;
    String phone;
    String email;

    ReturnedResult(String id, String phone, String email) {
      this.id = id;
      this.phone = phone;
      this.email = email;
    }

    String toJson() {
      JSONObject object = new JSONObject();
      try {
        object.put("phone", phone);
        if (email != null) {
          object.put("email", email);
        }
        if (id != null) {
          object.put("id", id);
        }
      } catch (JSONException e) {
        e.printStackTrace();
      }
      return object.toString();
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';
import 'package:flutter/services.dart';
import 'package:fb_accountkit_flutter/fb_accountkit_flutter.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  Result _loginResult = Result();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Account Kit Flutter example app'),
          backgroundColor: Colors.redAccent,
        ),
        body: Center(
          child: Column(
            mainAxisSize: MainAxisSize.max,
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              RaisedButton(
                  onPressed: () {
                    openFacebookActivity();
                  },
                  child: Text("Open Account Kit SMS Authenation")),
              Text('Result:  ${_loginResult.message == null ? "Ready to Process" : _loginResult.message}\n'),
              _getOptionalInformation()
            ],
          ),
        ),
      ),
    );
  }

  Future<void> openFacebookActivity() async {
    Result result;

    try {

      FbAccountkitFlutter.defaultCountryCode = "[YOUR DEFAULT COUNTRY CODE]"; // Ex:  US, AL etc.
      // Find more here : https://developers.facebook.com/docs/accountkit/countrycodes/

      FbAccountkitFlutter.codeIso = "[YOUR ISO CODE]"; // Ex: 93, 355 etc.
      // Find Dialing Code here : https://developers.facebook.com/docs/accountkit/countrycodes/

      FbAccountkitFlutter.countryCode = "[YOUR COUNTRY CODE]";
      // Find A 3 in here : http://kirste.userpage.fu-berlin.de/diverse/doc/ISO_3166.html

      // Optional
      FbAccountkitFlutter.phone = "[INITIAL PHONE]";

      // Request open accountkit
      result = await FbAccountkitFlutter.startAuthentication();

    } on PlatformException catch(e) {
      print(e);
      result = null;
    }

    setState(() {
      _loginResult = result;
    });
  }

  Widget _getOptionalInformation() {
    var phone = _loginResult.phone == null ? "Phone Null" : _loginResult.phone;
    var email = _loginResult.email == null ? "Email Null" : _loginResult.email;
    var userId = _loginResult.userId == null ? "User Id Null" : _loginResult.userId;
    if (_loginResult.status == FbAccountkitFlutter.SUCCEED) {
      return Text("Optional Info: \n Phone -> $phone \n Email -> $email \n UserId -> $userId");
    }
    return Container();
  }
}
